RecoCTPPS/TotemRPLocal/src/TotemTimingConversions.cc

```diff
@@ -3,119 +3,157 @@
 {
+  if (!calibrationFileOpened_)
+    openCalibrationFile();
+
   unsigned int offsetOfSamples = digi.getEventInfo().getOffsetOfSamples();
   if (offsetOfSamples == 0)
     offsetOfSamples = 30; // FW 0 is not sending this, FW > 0 yes
 
-  float triggerCellTimeInstant = getTimeOfFirstSample(digi) +
-        (SAMPIC_MAX_NUMBER_OF_SAMPLES - offsetOfSamples) *
-        SAMPIC_SAMPLING_PERIOD_NS;
+  return getTimeOfFirstSample(digi) + (SAMPIC_MAX_NUMBER_OF_SAMPLES - offsetOfSamples) * SAMPIC_SAMPLING_PERIOD_NS;
+}
 
-  return triggerCellTimeInstant;
+//----------------------------------------------------------------------------------------------------
+
+float
+TotemTimingConversions::getTimePrecision(const TotemTimingDigi& digi)
+{
+  if (!calibrationFileOpened_)
+    openCalibrationFile();
+  int db = digi.getHardwareBoardId();
+  int sampic = digi.getHardwareSampicId();
+  int channel = digi.getHardwareChannelId();
+  return parsedData_.getTimePrecision(db, sampic, channel);
 }
 
 //----------------------------------------------------------------------------------------------------
 
-std::vector<float> TotemTimingConversions::getTimeSamples(const TotemTimingDigi& digi) const
+std::vector<float>
+TotemTimingConversions::getTimeSamples(const TotemTimingDigi& digi)
 {
+  if (!calibrationFileOpened_)
+    openCalibrationFile();
+
   std::vector<float> time(digi.getNumberOfSamples());
-  // firstCellTimeInstant = 0;
-  if (!calibrationFileOk_)
-  {
-    for (unsigned int i = 0; i < time.size(); ++i)
-      time.at(i) = getTimeOfFirstSample(digi) + i * SAMPIC_SAMPLING_PERIOD_NS;
-  }
+  for (unsigned int i = 0; i < time.size(); ++i)
+    time.at(i) = getTimeOfFirstSample(digi) + i * SAMPIC_SAMPLING_PERIOD_NS;
   return time;
 }
 
 //----------------------------------------------------------------------------------------------------
+// NOTE: If no proper file is specified, calibration is not applied
 
-std::vector<float> TotemTimingConversions::getVoltSamples(const TotemTimingDigi& digi) const
+std::vector<float>
+TotemTimingConversions::getVoltSamples(const TotemTimingDigi& digi)
 {
+  if (!calibrationFileOpened_)
+    openCalibrationFile();
+
   std::vector<float> data;
-  if (!calibrationFileOk_)
-  {
+  if (calibrationFile_.empty())
     for (auto it = digi.getSamplesBegin(); it != digi.getSamplesEnd(); ++it)
       data.emplace_back(SAMPIC_ADC_V * (*it));
+  else {
+    unsigned int db = digi.getHardwareBoardId();
+    unsigned int sampic = digi.getHardwareSampicId();
+    unsigned int channel = digi.getHardwareChannelId();
+    unsigned int cell = digi.getCellInfo();
+    for (auto it = digi.getSamplesBegin(); it != digi.getSamplesEnd(); ++it, ++cell){
+      auto parameters = parsedData_.getParameters(db, sampic, channel, cell);
+      for (unsigned int i=0; i<parameters.size(); ++i)
+        calibrationFunction_.SetParameter(i, parameters.at(i));
+      data.emplace_back(calibrationFunction_.Eval(*it));
```

Comment:
```data.emplace_back(calibrationFunction_.EvalPar(sample, parameters.data());```

Reply:
It looks like ```parsedData_``` will return an empty vector for a corrupt/invalid data.
Your current implementation will simply call ```Eval``` with the last valid parameters (perhaps from another detector). This doesn't look good and should be protected.

Reply:
Both suggestions implemented. Thanks!